trieste/acquisition/multi_objective/dominance.py

```diff
@@ -15,30 +15,46 @@
 given a set of data points."""
 from __future__ import annotations
 
+import numpy as np
 import tensorflow as tf
 
 from ...types import TensorType
+from ...utils import to_numpy
 
 
 def non_dominated(observations: TensorType) -> tuple[TensorType, TensorType]:
     """
-    Computes the non-dominated set for a set of data points.
+    Computes the non-dominated set for a set of data points. Based on:
+    https://stackoverflow.com/questions/32791911/fast-calculation-of-pareto-front-in-python
+
     If there are duplicate point(s) in the non-dominated set, this function will return
     as it is without removing the duplicate.
 
     :param observations: set of points with shape [N,D]
-    :return: tf.Tensor of the non-dominated set [P,D] and the degree of dominance [N],
-        P is the number of points in pareto front dominances gives the number of
-        dominating points for each data point
+    :return: tf.Tensor of the non-dominated set [P,D] and a non-dominated point mask [N],
+        P is the number of points in pareto front, the mask specifies whether each data point
+        is non-dominated or not.
     """
-    extended = tf.tile(observations[None], [len(observations), 1, 1])
-    swapped_ext = tf.transpose(extended, [1, 0, 2])
-    dominance = tf.math.count_nonzero(
-        tf.logical_and(
-            tf.reduce_all(extended <= swapped_ext, axis=2),
-            tf.reduce_any(extended < swapped_ext, axis=2),
-        ),
-        axis=1,
-    )
-
-    return tf.boolean_mask(observations, dominance == 0), dominance
+    orig_points = observations.shape[0]
+    orig_indices = np.arange(orig_points)
+
+    if orig_points > 0:
+        # Reordering the observations beforehand speeds up the search
+        obs_np = to_numpy(observations)
+        idx = np.argsort(((obs_np - obs_np.mean(axis=0)) / (obs_np.std(axis=0) + 1e-7)).sum(axis=1))
```

Comment:
will this trigger retracing if this function is used anywhere in tf.function?

Reply:
Very good spot! In fact the use of not just `to_numpy` but also `np.any` and `np.sum` in the main loop all fail inside a tf.function! I'll have a go at rewriting this in pure tensorflow.

Reply:
I've added a test, but fixing this will probably be a pain :(